src/common/linear_assignment_optimizer.h

```diff
@@ -156,3 +158,129 @@ void auctionOptimization(M &value_matrix, std::array<int, N> &assignment, int si
 }
 
 }  // linear_assignment
+
+class LinearAssignmentCache {
+public:
+	static constexpr LinearAssignmentCache* kNotProvidedCache = nullptr;
+
+	class OptimizerInputOutput {
+	public:
+		OptimizerInputOutput()
+		    : target_slice_(Goal::Slice::Type(1)),
+		      source_slice_(Goal::Slice::Type(1)),
+		      assignemt_() {}
+
+		template <std::size_t N>
+		bool checkMatchAndGetResult(Goal::Slice &target_slice,
+		                            Goal::Slice &source_slice,
+		                            std::array<int, N> &assignemt) {
+			if (target_slice.getType() != target_slice_.getType() ||
+			    source_slice.getType() != source_slice_.getType()) {
+				return false;
+			}
+			assert(target_slice.size() == target_slice_.size());
+			assert(source_slice.size() == source_slice_.size());
+
+			for (int i = 0; i < target_slice.size(); i++) {
+				detail::Slice::PartProxy part = target_slice[i];
+				detail::Slice::PartProxy part_ = target_slice_[i];
+				if (part.size() != part_.size()) { return false; }
+				auto part_it_ = part_.begin();
+				for (const auto &label : part) {
+					if (label != *part_it_) { return false; }
+					part_it_++;
+				}
+			}
+			for (int i = 0; i < source_slice.size(); i++) {
+				detail::Slice::PartProxy part = source_slice[i];
+				detail::Slice::PartProxy part_ = source_slice_[i];
+				if (part.size() != part_.size()) { return false; }
+				auto part_it_ = part_.begin();
+				for (const auto &label : part) {
+					if (label != *part_it_) { return false; }
+					part_it_++;
+				}
+			}
+
+			// match is good
+			memcpy(assignemt.data(), assignemt_.data(),
+			       sizeof(int) * target_slice.size());
+
+			return true;
+		}
+
+		template <std::size_t N>
+		void store(Goal::Slice &target_slice, Goal::Slice &source_slice,
+		           std::array<int, N> &assignemt) {
+			target_slice_ = target_slice;
+			source_slice_ = source_slice;
+			assignemt_.resize(target_slice_.size());
+			for (int i = 0; i < target_slice_.size(); i++) {
+				assignemt_[i] = assignemt[i];
+			}
+		}
+
+	protected:
+		Goal::Slice target_slice_, source_slice_;
+		std::vector<int> assignemt_;
+	};
+
+	using EntriesContainer = std::vector<OptimizerInputOutput>;
+	using HashType = int64_t;
+
+	LinearAssignmentCache() : cachedData_(kModSlice_) {}
+
+	HashType getLabelValue(MediaLabel label) {
+		return static_cast<MediaLabelManager::HandleValue>(label) ;
+	}
+
+	HashType HashPart(const detail::Slice::PartProxy &part) {
```

Comment:
why those functions are using capital letters ?
```suggestion
	HashType hashPart(const detail::Slice::PartProxy &part) {
```

Reply:
Applied.